### PR TITLE
P3M3: Non-cubic domains, fixes, improvements

### LINKED ISSUE #6997: Assertion failure with read_MEDIT
Original comment from stackoverflow:

*I have tested with the new function and the issue remains. The output is giving the error "Error in add_facet_to_incident_cells_map" before giving a segmentation fault. The .mesh file I am trying to load was generated by CGAL itself. I have made the mesh file available this [link](https://wetransfer.com/downloads/f095b06e268879212cc60c78bfdf5ea820221029064830/48c963)*

### LINKED ISSUE #7214: warning: possibly dangling reference to a temporary [-Wdangling-reference]
## Issue Details

In this [testsuite](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-166/Periodic_3_mesh_3/TestReport_lrineau_Fedora-rawhide.gz) we get a warning for this 
[line](https://github.com/CGAL/cgal/blob/master/Periodic_3_mesh_3/include/CGAL/Periodic_3_mesh_3/Protect_edges_sizing_field.h#L1033) we get the warning about a possibly dangling reference.  Although it is a false alert, is there any gain from `const&` here.
Won't there anyways be a  returned objects optimization.

### PR DESCRIPTION
**Pull request by MaelRL**
## Summary of Changes

- Add a dummy point generator that is agnostic to the periodic domain's geometry, allowing P3M3 to handle non-cubic domain.
- Add an example of P3M3 running on polyhedral domains.
- Fix a few issues in SMDS_3's tet soup reading relating to non-manifoldness + general IO code speed improvements.
- Fix a few issues in P3M3's medit output and always duplicate facets to reduce the quantity of non-manifoldness occurences.
- Fix an issue with including Mesh_3 headers before P3M3's.
- Fix periodic point projection (for Mesh_3 optimizers) badly computing the projection direction
- Robustify a handful of P3M3 functions that were using intermediate constructions rather than the <Point, Offset> API (which uses filtered predicates and thus yields proper robustness)
- Misc minor code cleaning / modernization

ToDo:
- [x] Check if some medit IO issues are fixed
- [x] Add non-manifoldness flag to read_MEDIT
- [x] Fix value of orthosphere_radius_threshold
- [x] Check calls of insert_dummy_points() (use generic)
- [x] Check doc of P3M3 about cubic domains requirement

Not done:
- [ ] Fix sharp features for periodic polyhedral domains (problems due to changing weights creating cover change...). Code: https://gist.github.com/MaelRL/357c52dc04e16a162f13a3c6860a1398; WIP diff: https://gist.github.com/MaelRL/f98809964acffbacb6d3f42793e2654e

## Release Management

* Affected package(s): `Mesh_3`,  `P3M3`, `P3T3`, `SMDS3`
* Issue(s) solved (if any): Fix #6997, Fix #7214
* Feature/Small Feature (if any): TODO
* License and copyright ownership: no change

### CODE REVIEW

**Comment by MaelRL** [comment]
I didn't notice any slowdown of pure Mesh_3 as a consequence of [`06d2721` (#7062)](https://github.com/CGAL/cgal/pull/7062/commits/06d272169feeeb6d4c0ad93d83f0624f1666becd).

**Comment by MaelRL** [comment]
It is not documented, but it appears in P3M3 examples, which seems to be a criterion to be added to `list_of_documented_headers`.

(Probably a sign that it should be documented properly, actually.)

**Comment by lrineau** [comment]
The relevant log is now:
```
In file included from /home/runner/work/cgal/cgal/Mesh_3/include/CGAL/Compact_mesh_cell_base_3.h:19,
                 from /home/runner/work/cgal/cgal/dep_check_build/test_headers.cpp:452:
/home/runner/work/cgal/cgal/Mesh_3/include/CGAL/Mesh_3/config.h:15:2: warning: #warning Mesh_3/config.h [-Wcpp]
   15 | #warning Mesh_3/config.h
      |  ^~~~~~~
In file included from /home/runner/work/cgal/cgal/Periodic_3_mesh_3/include/CGAL/Periodic_3_mesh_triangulation_3.h:20,
                 from /home/runner/work/cgal/cgal/Periodic_3_mesh_3/include/CGAL/Periodic_3_function_wrapper.h:18,
                 from /home/runner/work/cgal/cgal/dep_check_build/test_headers.cpp:577:
/home/runner/work/cgal/cgal/Periodic_3_mesh_3/include/CGAL/Periodic_3_mesh_3/config.h:24: error: #error "The header '<CGAL/Periodic_3_mesh_3/config.h>' must be included before '<CGAL/Mesh_3/config.h>'"
   24 |   #error "The header '<CGAL/Periodic_3_mesh_3/config.h>' must be included before '<CGAL/Mesh_3/config.h>'"
      | 
```

The generate `test_headers.cpp` includes all documented headers. The issue is that documented headers of Mesh_3 are included before documented headers of P3M3. I have fixed that issue by including `<CGAL/Periodic_3_mesh_3/config.h>` at the top of the file (b5e7ffdeea951250af77c8810abccff9bf750f6b).

**Comment by MaelRL** [comment]
> The generate test_headers.cpp includes all documented headers. The issue is that documented headers of Mesh_3 are included before documented headers of P3M3.

Yes that's what I meant with "M comes before P" in the OP.

> I have fixed that issue by including <CGAL/Periodic_3_mesh_3/config.h> at the top of the file 

I think it's better to whitelist it since P3M3 disable macros and thus code that main Mesh_3 uses and it's better if the CI tool checks Mesh_3's real compilation than the version that P3M3 uses.

**Comment by lrineau** [comment]
> > I have fixed that issue by including <CGAL/Periodic_3_mesh_3/config.h> at the top of the file
> 
> I think it's better to whitelist it since P3M3 disables macros and thus code that main Mesh_3 uses and it's better if the CI tool checks Mesh_3's real compilation than the version of Mesh_3 that P3M3 uses.

There is no way that a whitelisting can work, unless one whitelists all of `Mesh_3`, or all of `Periodic_3_mesh_3`, because the content of `test_headers.cpp` will still include `Mesh_3/config.h` before `Periodic_3_mesh_3/config.h`. The first header in that file to in include `Mesh_3/config.h` is `<CGAL/Compact_mesh_cell_base_3.h>`, and probably all headers of `Mesh_3` include it. Then, any header of `Periodic_3_mesh_3` will probably include `Periodic_3_mesh_3/config.h`.

The test of `test_headers.cpp` was created to check that headers from CGAL do not conflict with each other. The fact that `P3M3` slightly modifies `Mesh_3` is not really an issue, in my opinion. The only alternative would be to whitelist all headers of `Periodic_3_mesh_3`.

**Comment by sloriot** [comment]
New warnings in CGAL-5.6-Ic-144.
For example [here](https://cgal.geometryfactory.com/CGAL/Members/testsuite/CGAL-5.6-Ic-144/Alpha_shapes_3_Examples/TestReport_magritte_arm64-apple_bigsur_clang1200-release-64bits.gz), [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-144/Periodic_3_triangulation_3_Examples/TestReport_magritte_arm64-apple_bigsur_clang1200-release-64bits.gz).

`Periodic_3_triangulation_3` is fully red.

**Comment by sloriot** [comment]
Maybe [this one in `Mesh_3`](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-144/Mesh_3_Examples/TestReport_magritte_arm64-apple_bigsur_clang1200-release-64bits.gz) too.

**Comment by sloriot** [comment]
One remaining compilation error in P3T3 tests in [CGAL-5.6-Ic-148](https://cgal.geometryfactory.com/CGAL/testsuite/results-5.6-Ic-148.shtml)

File: Mesh_3/include/CGAL/Mesh_3/C3T3_helpers.h [hunk 1]
```diff
@@ -3479,11 +3476,12 @@ get_least_square_surface_plane(const Vertex_handle& v,
...
 
-   return std::make_pair(plane,
-     ref_facet.first->get_facet_surface_center(ref_facet.second));
-}
-
 
+  // The surface center of a facet might have an offset in periodic triangulations
+  const Bare_point& ref_facet_scp = ref_facet.first->get_facet_surface_center(ref_facet.second);
+  const Bare_point& ref_point = tr_.get_closest_point(cp(position), ref_facet_scp);
+  return std::make_pair(plane, ref_point);
```
[the remark this answers is not in the document]
Comment (MaelRL) [reply Mesh_3/include/CGAL/Mesh_3/C3T3_helpers.h:3483]: `Triangulation::get_closest_point(Point_3 p, Point_3 q)` returns the periodic copy of `q` that is closest to `p`. 

For Mesh_3, `q` itself is returned so so the code is equivalent to what there was before.

**Comment by sloriot** [comment]
Run time errors in CGAL-5.6-Ic-165, see [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-165/Periodic_3_mesh_3/TestReport_gimeno_Debian-testing-clang-main.gz) for example.

**Comment by sloriot** [comment]
Errors and warnings:
https://cgal.geometryfactory.com/CGAL/testsuite/results-5.6-Ic-168.shtml#Periodic_3_mesh_3

**Comment by sloriot** [comment]
Still failing in [CGAL-5.6-Ic-186](https://cgal.geometryfactory.com/CGAL/testsuite/results-5.6-Ic-186.shtml)

**Comment by sloriot** [comment]
New warnings and errors in CGAL-5.6-Ic-203

**Comment by sloriot** [comment]
Warnings on windows:
- https://cgal.geometryfactory.com/CGAL/Members/testsuite/CGAL-5.6-Ic-212/Periodic_3_mesh_3/TestReport_Christo_MSVC2015-Release-64bits.gz
- https://cgal.geometryfactory.com/CGAL/Members/testsuite/CGAL-5.6-Ic-212/Periodic_3_mesh_3/TestReport_Christo_MSVC-2022-Community-Release.gz
- https://cgal.geometryfactory.com/CGAL/Members/testsuite/CGAL-5.6-Ic-212/Periodic_3_mesh_3/TestReport_Christo_MSVC-2019-Community-Release.gz

**Comment by sloriot** [comment]
There are also runtime errors in CGAL-5.6-Ic-212

**Comment by sloriot** [comment]
Runtime error: https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.6-Ic-242/Periodic_3_mesh_3/TestReport_gimeno_ArchLinux-clang-Release.gz

**Comment by sloriot** [comment]
Error is gone in CGAL-5.6-Ic-261! Let's wait the end of the testsuite and we'll be able to merge that branch. :rocket:

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.6-Ic-261

**Pull request merged**